File: churn.go
package main

import (
	"flag"
	"fmt"
	"math/rand"
	"runtime"
	"sync"
	"sync/atomic"
	"time"
	"golang.org/x/text/language"
	"golang.org/x/text/message"
)

func churner(result chan uint64, start *sync.WaitGroup, spots [] chan struct {}) {
	s := rand.NewSource(time.Now().UnixNano())
	rng := rand.New(s)

	count := uint64(0)
	start.Wait()
	for true {

		sem := spots[ rng.Intn(100) % len(spots) ];
		sem <- (struct {}{})
		<- sem;

		count += 1
		if  clock_mode && atomic.LoadInt32(&stop) == 1 { break }
		if !clock_mode && count >= stop_count { break }
	}

	atomic.AddInt64(&threads_left, -1);
	result <- count
}

func main() {
	var spot_cnt int

	spot_cntOpt := flag.Int("s", 1, "Number of spots in the system")

	bench_init()

	spot_cnt = *spot_cntOpt

	threads_left = int64(nthreads)

	result := make(chan uint64)
	var wg sync.WaitGroup
	wg.Add(1)

	spots := make([] chan struct {}, spot_cnt)
	for i := range spots {
		spots[i] = make(chan struct {}, 1)
	}

	for i := 0; i < nthreads; i++ {
		go churner(result, &wg, spots)
	}
	fmt.Printf("Starting\n");
	atomic.StoreInt32(&stop, 0)
	start := time.Now()
	wg.Done();
	wait(start, true);

	atomic.StoreInt32(&stop, 1)
	end := time.Now()
	duration := end.Sub(start)

	fmt.Printf("\nDone\n")

	for atomic.LoadInt64(&threads_left) != 0 {
		for i := range spots {
			select {
			case <- spots[i]:
			default:
			}
			select {
			case spots[i] <- (struct {}{}):
			default:
			}
			runtime.Gosched()
		}
	}

	global_counter := uint64(0)
	for i := 0; i < nthreads; i++ {
		global_counter += <- result
	}

	p := message.NewPrinter(language.English)
	p.Printf("Duration (ms)        : %d\n", duration.Milliseconds())
	p.Printf("Number of processors : %d\n", nprocs);
	p.Printf("Number of threads    : %d\n", nthreads);
	p.Printf("Number of spots      : %d\n", spot_cnt);
	p.Printf("Total Operations(ops): %15d\n", global_counter);
	// p.Printf("Total blocks         : %15d\n", global_blocks);
	p.Printf("Ops per second       : %18.2f\n", float64(global_counter) / duration.Seconds());
	p.Printf("ns per ops           : %18.2f\n", float64(duration.Nanoseconds()) / float64(global_counter))
	p.Printf("Ops per threads      : %15d\n", global_counter / uint64(nthreads))
	p.Printf("Ops per procs        : %15d\n", global_counter / uint64(nprocs))
	p.Printf("Ops/sec/procs        : %18.2f\n", (float64(global_counter) / float64(nprocs)) / duration.Seconds())
	p.Printf("ns per ops/procs     : %18.2f\n", float64(duration.Nanoseconds()) / (float64(global_counter) / float64(nprocs)))
}

File: churn.rs
use std::sync::Arc;
use std::sync::atomic::Ordering;
use std::time::Instant;

use tokio::runtime::Builder;
use tokio::sync;
use tokio::time::{sleep,Duration};

use rand::Rng;

use clap::{Arg, App};
use num_format::{Locale, ToFormattedString};

#[path = "../bench.rs"]
mod bench;

// ==================================================
struct Churner {
	start: sync::Notify,
}

async fn churn_main( this: Arc<Churner>, spots: Arc<Vec<sync::Semaphore>>, exp: Arc<bench::BenchData>, skip_in: bool ) -> u64 {
	let mut skip = skip_in;
	this.start.notified().await;

	let mut count:u64 = 0;
	loop {
		let r : usize = rand::thread_rng().gen();

		let spot : &sync::Semaphore = &spots[r % spots.len()];
		if !skip { spot.add_permits(1); }
		spot.acquire().await.forget();
		skip = false;

		count += 1;
		if  exp.clock_mode && exp.stop.load(Ordering::Relaxed) { break; }
		if !exp.clock_mode && count >= exp.stop_count { break; }
	}

	exp.threads_left.fetch_sub(1, Ordering::SeqCst);
	count
}

// ==================================================
fn main() {
	let options = App::new("Churn Tokio")
		.args(&bench::args())
		.arg(Arg::with_name("nspots")  .short("s").long("spots")  .takes_value(true).default_value("1").help("Number of spots in the system"))
		.get_matches();

	let nthreads  = options.value_of("nthreads").unwrap().parse::<usize>().unwrap();
	let nprocs    = options.value_of("nprocs").unwrap().parse::<usize>().unwrap();
	let nspots    = options.value_of("nspots").unwrap().parse::<usize>().unwrap();

	let exp = Arc::new(bench::BenchData::new(options, nthreads, None));

	let s = (1000000 as u64).to_formatted_string(&Locale::en);
	assert_eq!(&s, "1,000,000");

	let spots : Arc<Vec<sync::Semaphore>> = Arc::new((0..nspots).map(|_| {
		sync::Semaphore::new(0)
	}).collect());

	let thddata : Vec<Arc<Churner>> = (0..nthreads).map(|_| {
		Arc::new(Churner{
			start: sync::Notify::new(),
		})
	}).collect();

	let mut global_counter :u64 = 0;
	let mut duration : std::time::Duration = std::time::Duration::from_secs(0);
	let runtime = Builder::new_multi_thread()
		.worker_threads(nprocs)
		.enable_all()
		.build()
		.unwrap();

	runtime.block_on(async {
		let threads: Vec<_> = (0..nthreads).map(|i| {
			tokio::spawn(churn_main(thddata[i].clone(), spots.clone(), exp.clone(), i < nspots))
		}).collect();
		println!("Starting");

		sleep(Duration::from_millis(100)).await;

		let start = Instant::now();

		for i in 0..nthreads {
			thddata[i].start.notify_one();
		}

		duration = exp.wait(&start).await;

		println!("\nDone");

		for i in 0..nspots {
			spots[i].add_permits(10000);
		}

		for t in threads {
			let c = t.await.unwrap();
			global_counter += c;
		}
	});

	println!("Duration (ms)        : {}", (duration.as_millis()).to_formatted_string(&Locale::en));
	println!("Number of processors : {}", (nprocs).to_formatted_string(&Locale::en));
	println!("Number of threads    : {}", (nthreads).to_formatted_string(&Locale::en));
	println!("Number of spots      : {}", "6");
	println!("Total Operations(ops): {:>15}", (global_counter).to_formatted_string(&Locale::en));
	println!("Ops per second       : {:>15}", (((global_counter as f64) / duration.as_secs() as f64) as u64).to_formatted_string(&Locale::en));
	println!("ns per ops           : {:>15}", ((duration.as_nanos() as f64 / global_counter as f64) as u64).to_formatted_string(&Locale::en));
	println!("Ops per threads      : {:>15}", (global_counter / nthreads as u64).to_formatted_string(&Locale::en));
	println!("Ops per procs        : {:>15}", (global_counter / nprocs as u64).to_formatted_string(&Locale::en));
	println!("Ops/sec/procs        : {:>15}", ((((global_counter as f64) / nprocs as f64) / duration.as_secs() as f64) as u64).to_formatted_string(&Locale::en));
	println!("ns per ops/procs     : {:>15}", ((duration.as_nanos() as f64 / (global_counter as f64 / nprocs as f64)) as u64).to_formatted_string(&Locale::en));
}
